In [2]:
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt

In [5]:
data1=pd.read_excel("data1.xlsx")
data1.head()

Order ID       Product ID  Qty
0  CA-2014-100006  TEC-PH-10002075    3
1  CA-2014-100090  FUR-TA-10003715    3
2  CA-2014-100090  OFF-BI-10001597    6
3  CA-2014-100293  OFF-PA-10000176    6
4  CA-2014-100328  OFF-BI-10000343    1

In [6]:
data2=pd.read_excel("data2.xlsx")
data2.head()

Order ID       Product ID  Qty
0  CA-2014-100090  FUR-TA-10003715    3
1  CA-2014-100090  OFF-BI-10001597    6
2  CA-2014-100293  OFF-PA-10000176    6
3  CA-2014-100328  OFF-BI-10000343    1
4  CA-2014-100363  OFF-FA-10000611    2

In [91]:
Record=data1.iloc[:,0:2]
Record

Order ID       Product ID
0     CA-2014-100006  TEC-PH-10002075
1     CA-2014-100090  FUR-TA-10003715
2     CA-2014-100090  OFF-BI-10001597
3     CA-2014-100293  OFF-PA-10000176
4     CA-2014-100328  OFF-BI-10000343
...              ...              ...
9473  US-2017-169551  OFF-PA-10004100
9474  US-2017-169551  OFF-ST-10004835
9475  US-2017-169551  TEC-AC-10002018
9476  US-2017-169551  TEC-AC-10003033
9477  US-2017-169551  TEC-PH-10001363

[9478 rows x 2 columns]

In [164]:
data1["Record1"]=data1["Order ID"]+data1["Product ID"]
data2["Record2"]=data2["Order ID"]+data2["Product ID"]

# How to identify the Records (Order ID + Product ID combination) present in data1 but missing in data2 (Specify the number of records missing in your answer)

In [165]:
question1 = ~data1["Record1"].isin(data2["Record2"])
rest1=data1.loc[question1, "Record1"]

In [166]:
question1

0        True
1       False
2       False
3       False
4       False
        ...  
9473    False
9474    False
9475    False
9476    False
9477    False
Name: Record1, Length: 9478, dtype: bool

In [167]:
rest1

0       CA-2014-100006TEC-PH-10002075
10      CA-2014-100678OFF-EN-10000056
19      CA-2014-100895OFF-AR-10004511
35      CA-2014-101560OFF-BI-10000309
61      CA-2014-102673OFF-LA-10001771
                    ...              
9390    US-2017-160836OFF-AP-10001626
9403    US-2017-162558FUR-FU-10002364
9420    US-2017-163657OFF-BI-10000138
9427    US-2017-164056FUR-TA-10001307
9435    US-2017-165456FUR-CH-10003981
Name: Record1, Length: 507, dtype: object

In [168]:
question1.value_counts()

False    8971
True      507
Name: Record1, dtype: int64

# ANSWER=507

# How to identify the Records (Order ID + Product ID combination) missing in data1 but present in data2 (Specify the number of records missing in your answer)

In [169]:
question2 = ~data2["Record2"].isin(data1["Record1"])
rest2=data2.loc[question2, "Record2"]

In [234]:
question2.value_counts()

False    8971
True      508
Name: Record2, dtype: int64

# ANSWER  IS 508

In [233]:
rest3=rest2.values
rest3

array(['CA-2014-100706TEC-AC-10001314', 'CA-2014-100762OFF-PA-10001815',
       'CA-2014-101427OFF-AR-10002257', 'CA-2014-102652FUR-FU-10001918',
       'CA-2014-102869OFF-PA-10000788', 'CA-2014-103310TEC-PH-10001817',
       'CA-2014-103331OFF-PA-10002160', 'CA-2014-103492OFF-BI-10004817',
       'CA-2014-104829OFF-PA-10003016', 'CA-2014-105165FUR-TA-10004154',
       'CA-2014-105648FUR-TA-10002958', 'CA-2014-106439OFF-AR-10001419',
       'CA-2014-106439TEC-AC-10004568', 'CA-2014-108861OFF-BI-10003876',
       'CA-2014-110611OFF-ST-10000642', 'CA-2014-111150OFF-AR-10000034',
       'CA-2014-111962OFF-BI-10001308', 'CA-2014-112158TEC-AC-10003628',
       'CA-2014-113166OFF-PA-10001947', 'CA-2014-113271OFF-AR-10003251',
       'CA-2014-113271OFF-PA-10002365', 'CA-2014-113383FUR-FU-10001986',
       'CA-2014-113383OFF-AP-10004532', 'CA-2014-114314OFF-LA-10000240',
       'CA-2014-114510OFF-ST-10003221', 'CA-2014-115056OFF-AP-10003971',
       'CA-2014-115259OFF-FA-10000621', 'CA-2014-11

In [172]:
data2["rest2"]=pd.DataFrame(rest3)
data2

Order ID       Product ID  Qty                         Record  \
0     CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090FUR-TA-10003715   
1     CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090OFF-BI-10001597   
2     CA-2014-100293  OFF-PA-10000176    6  CA-2014-100293OFF-PA-10000176   
3     CA-2014-100328  OFF-BI-10000343    1  CA-2014-100328OFF-BI-10000343   
4     CA-2014-100363  OFF-FA-10000611    2  CA-2014-100363OFF-FA-10000611   
...              ...              ...  ...                            ...   
9474  US-2017-169551  OFF-PA-10004100    3  US-2017-169551OFF-PA-10004100   
9475  US-2017-169551  OFF-ST-10004835    3  US-2017-169551OFF-ST-10004835   
9476  US-2017-169551  TEC-AC-10002018    3  US-2017-169551TEC-AC-10002018   
9477  US-2017-169551  TEC-AC-10003033    2  US-2017-169551TEC-AC-10003033   
9478  US-2017-169551  TEC-PH-10001363    2  US-2017-169551TEC-PH-10001363   

                              rest2                        Record2  
0     CA-2014-100706TEC-AC-10001314  CA-2014-100090FUR-TA-10003715  
1     CA-2014-100762OFF-PA-10001815  CA-2014-100090OFF-BI-10001597  
2     CA-2014-101427OFF-AR-10002257  CA-2014-100293OFF-PA-10000176  
3     CA-2014-102652FUR-FU-10001918  CA-2014-100328OFF-BI-10000343  
4     CA-2014-102869OFF-PA-10000788  CA-2014-100363OFF-FA-10000611  
...                             ...                            ...  
9474                            NaN  US-2017-169551OFF-PA-10004100  
9475                            NaN  US-2017-169551OFF-ST-10004835  
9476                            NaN  US-2017-169551TEC-AC-10002018  
9477                            NaN  US-2017-169551TEC-AC-10003033  
9478                            NaN  US-2017-169551TEC-PH-10001363  

[9479 rows x 6 columns]

In [173]:
rest4=rest1.values

In [174]:
data1["rest1"]=pd.DataFrame(rest4)
data1

Order ID       Product ID  Qty                         Record  \
0     CA-2014-100006  TEC-PH-10002075    3  CA-2014-100006TEC-PH-10002075   
1     CA-2014-100090  FUR-TA-10003715    3  CA-2014-100090FUR-TA-10003715   
2     CA-2014-100090  OFF-BI-10001597    6  CA-2014-100090OFF-BI-10001597   
3     CA-2014-100293  OFF-PA-10000176    6  CA-2014-100293OFF-PA-10000176   
4     CA-2014-100328  OFF-BI-10000343    1  CA-2014-100328OFF-BI-10000343   
...              ...              ...  ...                            ...   
9473  US-2017-169551  OFF-PA-10004100    3  US-2017-169551OFF-PA-10004100   
9474  US-2017-169551  OFF-ST-10004835    3  US-2017-169551OFF-ST-10004835   
9475  US-2017-169551  TEC-AC-10002018    3  US-2017-169551TEC-AC-10002018   
9476  US-2017-169551  TEC-AC-10003033    2  US-2017-169551TEC-AC-10003033   
9477  US-2017-169551  TEC-PH-10001363    2  US-2017-169551TEC-PH-10001363   

                              rest1                        Record1  
0     CA-2014-100006TEC-PH-10002075  CA-2014-100006TEC-PH-10002075  
1     CA-2014-100678OFF-EN-10000056  CA-2014-100090FUR-TA-10003715  
2     CA-2014-100895OFF-AR-10004511  CA-2014-100090OFF-BI-10001597  
3     CA-2014-101560OFF-BI-10000309  CA-2014-100293OFF-PA-10000176  
4     CA-2014-102673OFF-LA-10001771  CA-2014-100328OFF-BI-10000343  
...                             ...                            ...  
9473                            NaN  US-2017-169551OFF-PA-10004100  
9474                            NaN  US-2017-169551OFF-ST-10004835  
9475                            NaN  US-2017-169551TEC-AC-10002018  
9476                            NaN  US-2017-169551TEC-AC-10003033  
9477                            NaN  US-2017-169551TEC-PH-10001363  

[9478 rows x 6 columns]

In [177]:
final_data=data1.merge(data2,how='outer',left_on='Record1',right_on='Record2')
final_data

Order ID_x     Product ID_x  Qty_x                       Record_x  \
0     CA-2014-100006  TEC-PH-10002075    3.0  CA-2014-100006TEC-PH-10002075   
1     CA-2014-100090  FUR-TA-10003715    3.0  CA-2014-100090FUR-TA-10003715   
2     CA-2014-100090  OFF-BI-10001597    6.0  CA-2014-100090OFF-BI-10001597   
3     CA-2014-100293  OFF-PA-10000176    6.0  CA-2014-100293OFF-PA-10000176   
4     CA-2014-100328  OFF-BI-10000343    1.0  CA-2014-100328OFF-BI-10000343   
...              ...              ...    ...                            ...   
9981             NaN              NaN    NaN                            NaN   
9982             NaN              NaN    NaN                            NaN   
9983             NaN              NaN    NaN                            NaN   
9984             NaN              NaN    NaN                            NaN   
9985             NaN              NaN    NaN                            NaN   

                              rest1                        Record1  \
0     CA-2014-100006TEC-PH-10002075  CA-2014-100006TEC-PH-10002075   
1     CA-2014-100678OFF-EN-10000056  CA-2014-100090FUR-TA-10003715   
2     CA-2014-100895OFF-AR-10004511  CA-2014-100090OFF-BI-10001597   
3     CA-2014-101560OFF-BI-10000309  CA-2014-100293OFF-PA-10000176   
4     CA-2014-102673OFF-LA-10001771  CA-2014-100328OFF-BI-10000343   
...                             ...                            ...   
9981                            NaN                            NaN   
9982                            NaN                            NaN   
9983                            NaN                            NaN   
9984                            NaN                            NaN   
9985                            NaN                            NaN   

          Order ID_y     Product ID_y  Qty_y                       Record_y  \
0                NaN              NaN    NaN                            NaN   
1     CA-2014-100090  FUR-TA-10003715    3.0  CA-2014-100090FUR-TA-10003715   
2     CA-2014-100090  OFF-BI-10001597    6.0  CA-2014-100090OFF-BI-10001597   
3     CA-2014-100293  OFF-PA-10000176    6.0  CA-2014-100293OFF-PA-10000176   
4     CA-2014-100328  OFF-BI-10000343    1.0  CA-2014-100328OFF-BI-10000343   
...              ...              ...    ...                            ...   
9981  US-2017-165344  OFF-BI-10003196   10.0  US-2017-165344OFF-BI-10003196   
9982  US-2017-165358  TEC-CO-10001943    5.0  US-2017-165358TEC-CO-10001943   
9983  US-2017-167920  OFF-AP-10000159    5.0  US-2017-167920OFF-AP-10000159   
9984  US-2017-169502  OFF-AP-10001947    5.0  US-2017-169502OFF-AP-10001947   
9985  US-2017-169551  FUR-BO-10001519    3.0  US-2017-169551FUR-BO-10001519   

                              rest2                        Record2  
0                               NaN                            NaN  
1     CA-2014-100706TEC-AC-10001314  CA-2014-100090FUR-TA-10003715  
2     CA-2014-100762OFF-PA-10001815  CA-2014-100090OFF-BI-10001597  
3     CA-2014-101427OFF-AR-10002257  CA-2014-100293OFF-PA-10000176  
4     CA-2014-102652FUR-FU-10001918  CA-2014-100328OFF-BI-10000343  
...                             ...                            ...  
9981                            NaN  US-2017-165344OFF-BI-10003196  
9982                            NaN  US-2017-165358TEC-CO-10001943  
9983                            NaN  US-2017-167920OFF-AP-10000159  
9984                            NaN  US-2017-169502OFF-AP-10001947  
9985                            NaN  US-2017-169551FUR-BO-10001519  

[9986 rows x 12 columns]

In [123]:
question2.value_counts()

False    8971
True      508
Name: Record, dtype: int64

In [16]:
data_final=data1.merge(data2.drop_duplicates(),on=['Order ID','Product ID'],how='left',indicator=True)

In [20]:
data_final._merge.value_counts()

both          8971
left_only      507
right_only       0
Name: _merge, dtype: int64

# Find the Sum of the total Qty of Records missing in data1 but present in data2

In [224]:
sum_of_Qty2=final_data[final_data["Record1"].isna()]

In [225]:
sum_of_Qty2['Qty_y'].sum()

1956.0

# ANSWER IS 1956

# Find the total number of unique records (Order ID + Product ID combination) present in the combined dataset of data1 and data2

In [232]:
final_data.shape

(9986, 12)

# ANSWER IS 9986